qubes/tests/integ/vm_qrexec_gui.py

```diff
@@ -535,53 +534,79 @@ def test_221_audio_record_muted(self):
         local_user = grp.getgrnam('qubes').gr_mem[0]
-        record = self.loop.run_until_complete(
-            self.testvm1.run('parecord --raw audio_rec.raw'))
+        record = self.loop.run_until_complete(self.testvm1.run(
+                'parecord --raw --format=float32le --rate=44100 \
+                            --channels=1 audio_rec.raw'))
         # give it time to start recording
-        self.loop.run_until_complete(asyncio.sleep(0.5))
         p = subprocess.Popen(['sudo', '-E', '-u', local_user,
-            'paplay', '--raw'],
+            'paplay', '--raw', '--format=float32le',
+                    '--rate=44100', '--channels=1'],
             stdin=subprocess.PIPE)
-        p.communicate(audio_in)
-        # wait for possible parecord buffering
-        self.loop.run_until_complete(asyncio.sleep(1))
+        p.communicate(audio_in.astype(np.float32).tobytes())
         self.loop.run_until_complete(
             self.testvm1.run_for_stdio('pkill parecord || :'))
         _, record_stderr = self.loop.run_until_complete(record.communicate())
         if record_stderr:
             self.fail('parecord printed something on stderr: {}'.format(
                 record_stderr))
+
         recorded_audio, _ = self.loop.run_until_complete(
             self.testvm1.run_for_stdio('cat audio_rec.raw'))
-        # allow up to 20ms to be missing
-        if audio_in[:-3528] not in recorded_audio:
-            found_bytes = recorded_audio.count(audio_in[0])
-            all_bytes = len(audio_in)
+        rec = np.fromstring(recorded_audio, dtype=np.float32)
+        crossings = np.nonzero((rec[1:] > 0) & (rec[:-1] < 0))[0]
+        np.seterr('raise')
+        rec_freq = len(rec)/np.mean(np.diff(crossings))
+        if not sfreq*0.8 < rec_freq < sfreq*1.2:
             self.fail('VM not recorded expected data, '
-                      'missing {} bytes out of {}'.format(
-                          all_bytes-found_bytes, all_bytes))
+                    'frequency {} not in specified range'
+                    .format(rec_freq))
+
+    @unittest.skipUnless(spawn.find_executable('parecord'),
+                         "pulseaudio-utils not installed in dom0")
+    def test_220_audio_play(self):
+        self.common_audio_playback()
+
+    @unittest.skipUnless(spawn.find_executable('parecord'),
+                         "pulseaudio-utils not installed in dom0")
+    def test_221_audio_rec_muted(self):
+        self.common_audio_record_muted()
+
+    @unittest.skipUnless(spawn.find_executable('parecord'),
+                         "pulseaudio-utils not installed in dom0")
+    def test_222_audio_rec_unmuted(self):
+        self.common_audio_record_unmuted()
+
+    @unittest.skipUnless(spawn.find_executable('parecord'),
+                         "pulseaudio-utils not installed in dom0")
+    def test_223_audio_play_hvm(self):
+        self.testvm1.virt_mode = 'hvm'
+        self.testvm1.features['audio-model'] = 'ich6'
```

Comment:
I've tested it on a system without the stubdomain changes included and it still passed. I think some extra step is needed to verify if it really uses stubdomain audio. Or do you think core-admin-client change connecting `pacat-simple-vchan` to stubdomain only is enough?

Another thing is, if the `module-vchan-sink` is still loaded in the VM, it will create an output even if dom0/sys-audio part is not running. And it will freeze the playback until `pacat-simple-vchan` is started. Perhaps by coincidence the emulated audio card is chosen by default, but this sounds like a risky approach to rely on it. Better set the default audio sink explicitly. And while doing this, you can also check if emulated card is detected, as a preliminary test (this would solve the issue above too).

Reply:
Don't you mind if I just remove guest' module-vchan-sink in corresponding hvm tests? (pacmd unload-module)

Reply:
That should work too.